src/protocol.cpp

```diff
@@ -1924,6 +1995,65 @@ bool CProtocol::EvaluateCLReqConnClientsListMes ( const CHostAddress& InetAddr )
     return false; // no error
 }
 
+void CProtocol::CreateCLChannelLevelListMes  ( const CHostAddress&      InetAddr,
+                                               const CVector<uint16_t>& vecLevelList,
+                                               const int                iNumClients )
+{
+    // This must be a multiple of bytes at four bits per client
+    const int        iNumBytes = ( iNumClients + 1 ) / 2;
+    CVector<uint8_t> vecData( iNumBytes );
+    int              iPos = 0; // init position pointer
+
+    for ( int i = 0, j = 0; i < iNumClients; i += 2 /* pack two per byte */, j++ )
+    {
+        uint16_t levelLo = vecLevelList[i] & 0x0F;
+        uint16_t levelHi = ( i + 1 < iNumClients ) ? vecLevelList[i + 1] & 0x0F : 0x0F;
+        uint8_t  byte    = static_cast<uint8_t> ( levelLo | ( levelHi << 4 ) );
+
+        PutValOnStream ( vecData, iPos,
+            static_cast<uint32_t> ( byte ), 1 );
+    }
+
+    CreateAndImmSendConLessMessage ( PROTMESSID_CLM_CHANNEL_LEVEL_LIST,
+                                     vecData,
+                                     InetAddr );
+}
+
+bool CProtocol::EvaluateCLChannelLevelListMes  ( const CHostAddress&     InetAddr,
+                                                 const CVector<uint8_t>& vecData )
+{
+    int       iPos     = 0; // init position pointer
+    const int iDataLen = vecData.Size();  // four bits per channel, 2 channels per byte
```

Comment:
You could introduce a plausability check here. The maximum possible number of channels at the server is given in the global.h file. You could check that the length of the data vector does not exceed that number, i.e. the number of bits times the maximum possible number of channels.

Reply:
Yep, I'll add that (just before rebasing).